source/game_sa/Ropes.cpp

```diff
@@ -1,58 +1,125 @@
+        }
+    }
 }
-int32 CRopes::FindRope(uint32 id)
-{
-    return ((int32(__cdecl*)(uint32))0x556000)(id);
+
+// 0x558D10
+int CRopes::CreateRopeForSwatPed(const CVector& startPos) {
+//    auto mi = ModelIndices[1];
+    int32 newRopeId = m_nRopeIdCreationCounter + 100;
+//    bool ropeCreated = RegisterRope(newRopeId, 8, startPos, true, 0, 0, nullptr, 4000);
+//    if (!ropeCreated) {
+//        return -1;
+//    }
+
+    m_nRopeIdCreationCounter += 1;
+    return newRopeId;
 }
-void CRopes::Init()
-{
-    ((void(__cdecl*)())0x555DC0)();
+
+//0x556760
+float CRopes::FindPickupHeight(CEntity* entity) {
+    return CModelInfo::GetModelInfo(entity->m_nModelIndex)->GetColModel()->m_boundBox.m_vecMax.z;
 }
-bool CRopes::IsCarriedByRope(CEntity* entity)
-{
-    return ((bool(__cdecl*)(CEntity*))0x555F80)(entity);
+
+// 0x556000
+int CRopes::FindRope(uint32 id) {
+    for (auto ropeId = 0; ropeId < m_nNumRopes; ropeId++) {
+        if (aRopes[ropeId].m_nId == id) {
+            return ropeId;
+        }
+    }
+
+    return -1;
 }
-bool CRopes::RegisterRope(CEntity* ropeId, uint32 ropeType, CVector startPos, bool bExpires, uint8 segmentCount, uint8 flags, CEntity* holder, uint32 timeExpire)
-{
-    return ((bool(__cdecl*)(CEntity*, uint32, CVector, bool, uint8, uint8, CEntity*, uint32))0x556B40)(ropeId, ropeType, startPos, bExpires, segmentCount, flags, holder, timeExpire);
+
+// 0x555F80
+bool CRopes::IsCarriedByRope(CEntity* entity) {
+    if (!entity) {
+        return false;
+    }
+
+    for (auto ropeId = 0; ropeId < m_nNumRopes; ropeId++) {
+        if (aRopes[ropeId].m_pAttachedEntity == entity) {
+            return true;
+        }
+    }
+
+    return false;
 }
-void CRopes::Render()
-{
-    ((void(__cdecl*)())0x556AE0)();
+
+bool CRopes::RegisterRope(CEntity* ropeObj, uint32 ropeType, CVector startPos, bool bExpires, uint8 segmentCount, uint8 flags, CEntity* holder, uint32 timeExpire) {
+    return ((bool(__cdecl*)(CEntity*, uint32, CVector, bool, uint8, uint8, CEntity*, uint32))0x556B40)(ropeObj, ropeType, startPos, bExpires, segmentCount, flags, holder, timeExpire);
 }
-void CRopes::SetSpeedOfTopNode(uint32 ropeId, CVector dirSpeed)
-{
-    ((void(__cdecl*)(uint32, CVector))0x555DF0)(ropeId, dirSpeed);
+
+// 0x556AE0
+void CRopes::Render() {
+    for (auto ropeId = 0; ropeId < m_nNumRopes; ropeId++) {
+        aRopes[ropeId].Render();
+    }
 }
-void CRopes::Shutdown()
-{
-    ((void(__cdecl*)())0x556B10)();
+
+// 0x555DF0
+void CRopes::SetSpeedOfTopNode(uint32 ropeId, CVector dirSpeed) {
+  for (auto ropeIndex = 0; auto& rope : aRopes) {
+      if (rope.m_nType == eRopeType::CRANE_MAGNET1 || rope.m_nId != ropeId)
+          continue;
+
+      aRopes[ropeIndex].m_aSegmentsReleased[0] = dirSpeed;
+  }
 }
-void CRopes::Update()
-{
-    ((void(__cdecl*)())0x558D70)();
+
+// 0x558D70
+void CRopes::Update() {
+    if (CReplay::Mode == REPLAY_MODE_1) {
+        return;
+    }
+
+    for (uint8 ropeId = 0; ropeId < m_nNumRopes; ropeId++) {
```

Comment:
```suggestion
    for (auto ropeId = 0u; ropeId < m_nNumRopes; ropeId++) {
```
Or could use `uint32`.